src/trading/Trade.sol

```diff
@@ -0,0 +1,78 @@
+// Copyright (C) 2015 Forecast Foundation OU, full GPL notice in LICENSE
+
+pragma solidity ^0.4.13;
+
+import 'ROOT/Controlled.sol';
+import 'ROOT/libraries/ReentrancyGuard.sol';
+import 'ROOT/reporting/IMarket.sol';
+import 'ROOT/trading/IOrders.sol';
+import 'ROOT/trading/Trading.sol';
+
+
+// FIXME: Remove this once MakeOrder is implemented in Solidity
+contract IMakeOrder {
+    function makeOrder(address, Trading.TradeDirections, uint256, int256, IMarket, uint8, bytes32, bytes32, uint256) returns (bytes32);
+}
+
+
+// FIXME: Remove this once TakeOrder is implemented in Solidity
+contract ITakeOrder {
+    function takeOrder(address, bytes32, Trading.TradeTypes, IMarket, uint8, uint256, uint256) returns (uint256);
+}
+
+
+contract Trade is Controlled, ReentrancyGuard {
+    uint256 private constant MINIMUM_GAS_NEEDED = 300000;
+
+    function publicBuy(IMarket _market, uint8 _outcome, uint256 _fxpAmount, int256 _fxpPrice, uint256 _tradeGroupID) external onlyInGoodTimes nonReentrant returns (bytes32) {
+        bytes32 _output = this.trade(msg.sender, Trading.TradeDirections.Buying, _market, _outcome, _fxpAmount, _fxpPrice, _tradeGroupID);
+        return _output;
+    }
+
+    function publicSell(IMarket _market, uint8 _outcome, uint256 _fxpAmount, int256 _fxpPrice, uint256 _tradeGroupID) external onlyInGoodTimes nonReentrant returns (bytes32) {
+        bytes32 _output = this.trade(msg.sender, Trading.TradeDirections.Selling, _market, _outcome, _fxpAmount, _fxpPrice, _tradeGroupID);
+        return _output;
+    }
+
+    function publicTrade(Trading.TradeDirections _direction, IMarket _market, uint8 _outcome, uint256 _fxpAmount, int256 _fxpPrice, uint256 _tradeGroupID) external onlyInGoodTimes nonReentrant returns (bytes32) {
+        bytes32 _output = this.trade(msg.sender, _direction, _market, _outcome, _fxpAmount, _fxpPrice, _tradeGroupID);
+        return _output;
+    }
+
+    function publicTakeBestOrder(Trading.TradeDirections _direction, IMarket _market, uint8 _outcome, uint256 _fxpAmount, int256 _fxpPrice, uint256 _tradeGroupID) external onlyInGoodTimes nonReentrant returns (uint256) {
+        uint256 _fxpAmountRemaining = this.takeBestOrder(msg.sender, _direction, _market, _outcome, _fxpAmount, _fxpPrice, _tradeGroupID);
+        return _fxpAmountRemaining;
+    }
+
+    function trade(address _sender, Trading.TradeDirections _direction, IMarket _market, uint8 _outcome, uint256 _fxpAmount, int256 _fxpPrice, uint256 _tradeGroupID) external onlyWhitelistedCallers returns (bytes32) {
+        uint256 _bestFxpAmount = this.takeBestOrder(_sender, _direction, _market, _outcome, _fxpAmount, _fxpPrice, _tradeGroupID);
+        if (_bestFxpAmount > 0 && msg.gas >= MINIMUM_GAS_NEEDED) {
+            return IMakeOrder(controller.lookup("makeOrder")).makeOrder(_sender, _direction, _bestFxpAmount, _fxpPrice, _market, _outcome, 0, 0, _tradeGroupID);
+        }
```

Comment:
Style: Consider inverting the if statement here so this turns into a guard clause.  This case is a bit borderline, but my general philosophy is to have the "most interesting/most common code path" be least indented, and indents for side paths (less interesting or less common).  In this case, exiting early is not particularly interesting, but making an order is.

Also, consider breaking this into multiple simpler if statements like:
```
uint256 _bestFxpAmount = takeBestOrder(...);
if (_bestFxpAmount == 0) {
  return 1;
}
if (msg.gas < MINIMUM_GAS_NEEDED) {
  return 1;
}
return IMakeOrder(...).makeOrder(...);
```